primitives/xcm/src/barriers.rs

```diff
@@ -77,3 +77,97 @@ impl<T: Contains<MultiLocation>> ShouldExecute for AllowTopLevelPaidExecutionDes
 		}
 	}
 }
+
+/// Make sure that not withdrawable assets are handle properly according to the XCM design:
+/// - forbid any ReserveAssetDeposited instruction that contains not withdrawable asset(s) as
+/// parameter.
+/// - Morph some specific messages patterns that try to Widraw then Deposit not withdrawable
+/// asset(s).
+pub struct NotWithdrawableAssetsBarrier<IsWithdrawable, InnerBarrier>(
+	core::marker::PhantomData<(IsWithdrawable, InnerBarrier)>,
+);
+
+impl<IsWithdrawable: crate::IsWithdrawable, InnerBarrier: ShouldExecute> ShouldExecute
+	for NotWithdrawableAssetsBarrier<IsWithdrawable, InnerBarrier>
+{
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		instructions: &mut Xcm<Call>,
+		max_weight: Weight,
+		weight_credit: &mut Weight,
+	) -> Result<(), ()> {
+		InnerBarrier::should_execute(origin, instructions, max_weight, weight_credit)?;
+		let maybe_morphed_message = match &instructions.0[..] {
+			&[WithdrawAsset(ref assets), ClearOrigin, BuyExecution {
+				ref fees,
+				ref weight_limit,
+			}, DepositAsset {
+				assets: ref filter,
+				ref beneficiary,
+				..
+			}] => morph_xtokens_to_reserve_message::<Call, IsWithdrawable>(
+				assets,
+				fees,
+				weight_limit,
+				filter,
+				beneficiary,
+			),
+			_ => None,
+		};
+
+		if let Some(mut morphed_message) = maybe_morphed_message {
+			for i in 0..morphed_message.len() {
+				core::mem::swap(&mut instructions.0[i], &mut morphed_message[i]);
+			}
+		}
+
+		if instructions.0.iter().any(|instruction| {
+			matches!(
+				instruction, ReserveAssetDeposited(assets) if assets.inner().iter()
```

Comment:
Why do we need this? We cannot receive ReserveAssetDeposited from a token that is our reserve no (forbidden by IsReserve in XcmConfig?

Reply:
A local XCM run may well pass the `IsReserve` check

Reply:
AFAIK, not in our runtime. `IsReserve` filter everything where the reserve of the local asset != origin. In this case origin would be a local account, while the reserve would be `MutliLocation::here`